### PR TITLE
check rdma configuration and fix some logic problem

### PR DESCRIPTION
**Pull request by iotcg**
1. check rdma configuration is under hardware limitation.
2. fix ibv_port_attr object memory leak by using the object instead of allocating in the heap.
3. fix logic between RDMAV_HUGEPAGES_SAFE and ibv_fork_init.
4. fix error argument to get right qp state
5. separate Device construction when rdma_cm is used.
6. refine/simplify some function implementation.
7. decouple RDMAWorker & RDMAStack, RDMADispatcher & RDMAStack
8. remove redundant code.
9. rename var to improve readability.

Signed-off-by: Changcheng Liu <changcheng.liu@aliyun.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
[OT] @iotcg could you help review #25447? i briefly looked at it and it seems reasonable. but i am far from an expert in that area. and that PR definitely needs another pair of eyes.

**Comment by iotcg** [comment]
@tchaikov 
I'm going through async/RDMA code to enable new Network Interface Card. I‘ll check https://github.com/ceph/ceph/pull/25447 later.

File: src/msg/async/rdma/Infiniband.cc [hunk 1]
```diff
@@ -524,28 +524,33 @@ void Infiniband::MemoryManager::Chunk::set_bound(uint32_t b)
   bound = b;
 }
 
-void Infiniband::MemoryManager::Chunk::prepare_read(uint32_t b)
+uint32_t Infiniband::MemoryManager::Chunk::get_bound()
```
Comment (tchaikov) [inline src/msg/async/rdma/Infiniband.cc]: mark this method `const`

File: src/msg/async/rdma/Infiniband.cc [hunk 2]
```diff
@@ -524,28 +524,33 @@ void Infiniband::MemoryManager::Chunk::set_bound(uint32_t b)
...
 {
-  offset = 0;
-  bound = b;
+  return bound;
 }
 
-uint32_t Infiniband::MemoryManager::Chunk::get_bound()
+uint32_t Infiniband::MemoryManager::Chunk::get_size()
```
Comment (tchaikov) [inline src/msg/async/rdma/Infiniband.cc]: ditto.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 3]
```diff
@@ -257,31 +257,29 @@ void RDMAConnectedSocketImpl::handle_connection() {
 
 ssize_t RDMAConnectedSocketImpl::read(char* buf, size_t len)
 {
+  ssize_t read_size = 0;
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: please define this variable where it is used.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 4]
```diff
@@ -257,31 +257,29 @@ void RDMAConnectedSocketImpl::handle_connection() {
...
-  if (0 == connected) {
-    ldout(cct, 1) << __func__ << " when ib not connected. len: " << len <<dendl;
+  } else if (0 == connected) {
+    ldout(cct, 1) << __func__ << " ib not connected. len: " << len << dendl;
     return -EAGAIN;
   }
-  ssize_t read = 0;
   if (!buffers.empty())
-    read = read_buffers(buf,len);
+    read_size = read_buffers(buf, len);
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: i'd suggest have a separated commit to do the variable renaming and have smaller commits instead of composing a single fat one with multiple changes.

Comment (iotcg) [reply src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: @tchaikov I rename the variable when solving problems in some functions. Is it feasible to submit one patch to only change the variable name?

Comment (tchaikov) [reply src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: as long as your changes are not dependent on each other, and the name changing one does improve the readability, i think it'd be fine.

**Comment by iotcg** [comment]
@tchaikov I'm trying to unify Infiniband::MemoryManager::Chunk(manage receive buffer) and Infiniband::MemoryManager::Cluster(send buffer) implementation. They're marked #TODO in the source code. It's really confusing about their relationship in the current data structure.

Some other thinking in my side:
1) Whether we should use poll method or async event method in ceph/RDMA implementation. As being shown in https://github.com/ceph/ceph/pull/25447, it's said that some hardware doesn't support async event.
2) Some hardware support SRQ(shared receive queue), but current ceph/RDMA implementation bind some code implementation with the hardware support SRQ(When disabling SRQ in ceph configure, they don't work). Whether we should remove this limitation? If so, what's the impact of performance?

**Comment by rouming** [comment]
How can I remove my own approve?  That was made by mistake, I expected to approve commit by commit, not everything at once.

File: src/msg/async/rdma/Infiniband.cc [hunk 5]
```diff
@@ -780,10 +773,16 @@ Infiniband::MemoryManager::MemoryManager(CephContext *c, Device *d, ProtectionDo
     // allocate initial rxbuf_pool size as 2 * receive_queue_len if
     // it's possible which could avoid expanding pool upon receiving
     // the first packet.
-    rxbuf_pool(&rxbuf_pool_ctx, sizeof(Chunk) + c->_conf->ms_async_rdma_buffer_size,
-              (c->_conf->ms_async_rdma_receive_buffers >= 2 * c->_conf->ms_async_rdma_receive_queue_len ||
-               c->_conf->ms_async_rdma_receive_buffers <= 0) ?
-               2 * c->_conf->ms_async_rdma_receive_queue_len : c->_conf->ms_async_rdma_receive_buffers)
+    rxbuf_pool(&rxbuf_pool_ctx,
+               // nrequest_size: chunk size
+               sizeof(Chunk) + cct->_conf->ms_async_rdma_buffer_size,
```
Comment (rouming) [inline src/msg/async/rdma/Infiniband.cc]: Could you please not to mix cosmetic change with a fix? It's very hard to follow.

And here, please, make a separate static function for all the calculations with normal ifs, because here ternary operator is definitely overused. Impossible to read.

File: src/msg/async/rdma/Infiniband.h [hunk 6]
```diff
@@ -99,31 +100,40 @@ class DeviceList {
...
-    if (device_list == NULL || num == 0) {
-      lderr(cct) << __func__ << " failed to get rdma device list.  " << cpp_strerror(errno) << dendl;
-      ceph_abort();
+  explicit DeviceList(CephContext *cct): device_list(nullptr), device_context_list(nullptr),
+                                         num(0), devices(nullptr) {
+    device_list = ibv_get_device_list(&num);
+    ceph_assert(device_list);
+    ceph_assert(num);
+    if (cct->_conf->ms_async_rdma_cm) {
+        device_context_list = rdma_get_devices(&num);
```
[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/Infiniband.h]: @rouming Their value should be same.

[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/Infiniband.h]: @rouming If it's needed, I could change it to be rdma_get_devices(NULL)

File: src/msg/async/rdma/Infiniband.cc [hunk 7]
```diff
@@ -777,15 +777,13 @@ void Infiniband::MemoryManager::PoolAllocator::free(char * const block)
...
-               c->_conf->ms_async_rdma_receive_buffers > 0 ?
-                  // if possible make initial pool size 2 * receive_queue_len
-                  // that way there will be no pool expansion upon receive of the
-                  // first packet.
-                  (c->_conf->ms_async_rdma_receive_buffers < 2 * c->_conf->ms_async_rdma_receive_queue_len ?
-                   c->_conf->ms_async_rdma_receive_buffers :  2 * c->_conf->ms_async_rdma_receive_queue_len) :
-                  // rx pool is infinite, we can set any initial size that we want
-                   2 * c->_conf->ms_async_rdma_receive_queue_len)
+              (c->_conf->ms_async_rdma_receive_buffers >= 2 * c->_conf->ms_async_rdma_receive_queue_len ||
+               c->_conf->ms_async_rdma_receive_buffers <= 0) ?
```
[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/Infiniband.cc]: @rouming Thanks for the suggestion. I'll define a function to do that.

File: src/msg/async/rdma/Infiniband.cc [hunk 8]
```diff
@@ -1019,50 +1019,58 @@ Infiniband::QueuePair* Infiniband::create_queue_pair(CephContext *cct, Completio
...
-      if (i == 0)
+      lderr(cct) << __func__ << " WARNING: out of memory. Request " << rq_wr_num <<
+                 " rx buffers. Only get " << req_index << " rx buffers." << dendl;
+      if (req_index == 0) {
         return 0;
-      // if we got some buffers post them and hope for the best
-      rx_work_request[i-1].next = 0;
-      break;
+      } else {
+        break; //get some buffers, so we need post them to recevie queue
```
[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/Infiniband.cc]: @rouming 
1. I'll remove the "cosmetic change" i.e. keep "break" there.
2.  For the logic problem, I read it again. You're right. The previous object's next filed is set to be NULL when chunk == NULL;
3. For head/stack allocating, I just think it's not proper to allocate 4096 objects in the stack. But it's OK to do that in the stack. @rouming Which one do you prefer?
4. "num" is the parameter name, not the iterator.

@rouming & @tchaikov . Actually, this patch is submitted based on the idea to fix the "logic problem". Since the "logic problem" doesn't exist now, do I still need to refine this patch to only improve the readiness? Or, abandon this patch directly?

File: src/msg/async/rdma/Infiniband.cc [hunk 9]
```diff
@@ -175,7 +175,7 @@ Infiniband::QueuePair::QueuePair(
   q_key(q_key),
   dead(false)
 {
-  initial_psn = lrand48() & 0xffffff;
+  initial_psn = lrand48() & 0xffffffff;
```
[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/Infiniband.cc]: @rouming Thanks. I'm confusing about the bit number, it's clear now.

File: src/msg/async/rdma/Infiniband.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/msg/async/rdma/Infiniband.cc]: @iotcg i don't feel strong either way. but it's always nice to improve the readability.

File: src/msg/async/rdma/Infiniband.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/msg/async/rdma/Infiniband.cc]: probably we could take this chance to document this magic number with the InfiniBand spec?

File: src/msg/async/rdma/Infiniband.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rouming) [reply src/msg/async/rdma/Infiniband.h]: > @rouming Their value should be same.

That's correct, rdma_get_devices() uses ibv_get_device_list() internally and gets the same `num`. But, if you think about barely possible case, like loading the driver for some infiniband hardware between the calls (say you have mellanox card *and* qlogic, but for qlogic drivers are not loaded, so system does not see this hardware):  ibv_get_device_list() and rdma_get_devices(), then you get exactly this problem which I described.

File: src/msg/async/rdma/Infiniband.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (rouming) [reply src/msg/async/rdma/Infiniband.cc]: You did the right thing with your cleanups, just change the comment, because what you've described as a fix actually cleanups.

Regarding the stack: keep your variant. This value indeed eventually comes from the hardware, so we do not control it.

File: src/msg/async/rdma/Infiniband.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (rouming) [reply src/msg/async/rdma/Infiniband.cc]: Yes, Kefu is right, instead of removing the mask - add a short comment from the spec.

**Comment by rouming** [comment]
@iotcg  This is a good PR with some good catches, but I strongly suggest you to reorder all your patches and firstly have **only** one-line fixes, where it becomes immediately clear, what has been fixed exactly.  And then all cosmetics, renames, etc can follow. 

Also, please check the PR description.  IMO it does not correspond to what has been done here.

File: src/msg/async/rdma/Infiniband.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (rouming) [reply src/msg/async/rdma/Infiniband.cc]: > "num" is the parameter name, not the iterator.

Yeah, my comment regarding the 'i' iterator was about your other renaming, which you did not mention in the comment.

**Comment by iotcg** [comment]
> Need to refine: there's memory leak when return if i == 0.

Solved

**Comment by tchaikov** [comment]
@iotcg is this PR ready for another round of review? as @rouming has approved it. i will just walk through it to do some sanity checks if you are willing to rebase it and get it merged.

**Comment by iotcg** [comment]
Let me rebase the patches first.

**Comment by tchaikov** [comment]
retest this please.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 10]
```diff
@@ -618,11 +618,11 @@ void RDMAConnectedSocketImpl::cleanup() {
...
 void RDMAConnectedSocketImpl::notify()
 {
-  // note: notify_fd is an event fd (man eventfd)
-  // write argument must be a 64bit integer
-  uint64_t i = 1;
+  eventfd_t event_val = 1;
 
-  ceph_assert(sizeof(i) == write(notify_fd, &i, sizeof(i)));
+  int r = eventfd_write(notify_fd, event_val);
+
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nit, remove the empty line.

File: src/msg/async/rdma/RDMAStack.cc [hunk 11]
```diff
@@ -132,27 +132,98 @@ void RDMADispatcher::handle_async_event()
...
+              erase_qpn_lockless(qpn);
+          }
+        }
+        break;
+      case IBV_EVENT_QP_REQ_ERR:
+        /* Invalid Request Local Work Queue Error */
+        // fall through #TODO
+      case IBV_EVENT_QP_ACCESS_ERR:
+        /* Local access violation error */
+        // fall through #TODO
```
[the remark this answers is not in the document]
Comment (iotcg) [reply src/msg/async/rdma/RDMAStack.cc]: Thanks a lot. I'll use "[[fallthrough]];"

**Comment by iotcg** [comment]
retest this please.

**Comment by iotcg** [comment]
1. Follow Kefu's review comments to clean the code
2. Add one extra patch for helping debug RDMA initialization.
     https://github.com/ceph/ceph/pull/28344/commits/5c3aaae29be93cc91ef68e163fdc0fa236bfd799
3. Pass local build
4. Pass local test on Mellanox NIC under vstart.sh

**Comment by tchaikov** [comment]
@iotcg i added a prefix to your commit of "convert port_id from type uint8_t to int for output".

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-08-23_09:33:33-rados-wip-kefu-testing-2019-08-23-1454-distro-basic-mira/
